controllers/alert_controller.go

```diff
@@ -21,145 +21,185 @@ import (
 
-	return ctrl.Result{}, nil
-}
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
 
-func (r *AlertReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&v1beta1.Alert{}).
-		WithEventFilter(predicate.Or(predicate.GenerationChangedPredicate{}, predicates.ReconcileRequestedPredicate{})).
-		Complete(r)
+			readyCondition := conditions.Get(alert, meta.ReadyCondition)
+			switch readyCondition.Status {
+			case metav1.ConditionFalse:
+				// As we are no longer reconciling and the end-state is not ready, the reconciliation has stalled
+				conditions.MarkStalled(alert, readyCondition.Reason, readyCondition.Message)
+			case metav1.ConditionTrue:
+				// As we are no longer reconciling and the end-state is ready, the reconciliation is no longer stalled
+				conditions.Delete(alert, meta.StalledCondition)
+			}
+		}
+
+		if err := patchHelper.Patch(ctx, alert, patchOpts...); err != nil {
+			retErr = kerrors.NewAggregate([]error{retErr, err})
+		}
+
+		r.Metrics.RecordReadiness(ctx, alert)
+		r.Metrics.RecordDuration(ctx, alert, start)
+	}()
+
+	return r.reconcile(ctx, alert)
 }
 
-func (r *AlertReconciler) validate(ctx context.Context, alert v1beta1.Alert) error {
-	var provider v1beta1.Provider
-	providerName := types.NamespacedName{Namespace: alert.Namespace, Name: alert.Spec.ProviderRef.Name}
-	if err := r.Get(ctx, providerName, &provider); err != nil {
-		return fmt.Errorf("failed to get provider %s, error: %w", providerName.String(), err)
-	}
+func (r *AlertReconciler) reconcile(ctx context.Context, alert *v1beta1.Alert) (ctrl.Result, error) {
+	// Mark the resource as under reconciliation
+	conditions.MarkReconciling(alert, meta.ProgressingReason, "")
 
-	if !apimeta.IsStatusConditionTrue(provider.Status.Conditions, meta.ReadyCondition) {
-		return fmt.Errorf("provider %s is not ready", providerName.String())
+	// validate alert spec and provider
+	if err := r.validate(ctx, alert); err != nil {
+		conditions.MarkFalse(alert, meta.ReadyCondition, meta.FailedReason, err.Error())
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	return nil
+	conditions.MarkTrue(alert, meta.ReadyCondition, meta.SucceededReason, v1beta1.InitializedReason)
+	ctrl.LoggerFrom(ctx).Info("Alert initialized")
+
+	return ctrl.Result{}, nil
 }
 
-func (r *AlertReconciler) recordSuspension(ctx context.Context, alert v1beta1.Alert) {
-	if r.MetricsRecorder == nil {
-		return
+func (r *AlertReconciler) validate(ctx context.Context, alert *v1beta1.Alert) error {
+	provider := &v1beta1.Provider{}
+	providerName := types.NamespacedName{Namespace: alert.Namespace, Name: alert.Spec.ProviderRef.Name}
+	if err := r.Get(ctx, providerName, provider); err != nil {
+		// log not found errors since they get filtered out
+		ctrl.LoggerFrom(ctx).Error(err, "failed to get provider %s, error: %w", providerName.String())
+		return fmt.Errorf("failed to get provider '%s', error: %w", providerName.String(), err)
```

Comment:
```suggestion
		// log not found errors since they get filtered out
		ctrl.LoggerFrom(ctx).Error(err, "failed to get provider %s", providerName.String())
		return fmt.Errorf("failed to get provider '%s': %w", providerName.String(), err)
```

Using `, error:` is kind of redundant since Go has added support for proper error wrapping. In addition, the logger already takes `err` as an error, and the second argument sugars this error with an informational message.